File: MaxQueue/MaxQueue.cpp
class MaxQueue {
public:
    queue<int> q1;
    deque<int> q2;
    MaxQueue() {

    }
    
    int max_value() {
        int res = -1;
        if (!q2.empty())
        {
            res = q2.front();
        }
        return res;
    }
    
    void push_back(int value) {
        q1.push(value);
        while (!q2.empty() && q2.back()<value)
        {
            q2.pop_back();
        }
        q2.push_back(value);
    }
    
    int pop_front() {
        int res = -1;
        if (q1.empty()) return res;
        res = q1.front();
        if (res == q2.front())
        {
            q2.pop_front();
        }
        q1.pop();
        return res;
    }
};

File: MaxQueue/MaxQueue.py
class MaxQueue:

    def __init__(self):
        self.queue=[]
        self.max_queue=[]
        
    def max_value(self) -> int:
        if not self.queue:
            return -1
        return self.max_queue[0]
        

    def push_back(self, value: int) -> None:
        self.queue.append(value)
        while self.max_queue and value > self.max_queue[-1]:
                self.max_queue.pop(-1)
        self.max_queue.append(value)

    def pop_front(self) -> int:
        if not self.queue:
            return -1
        if self.queue[0]==self.max_queue[0]:
            self.max_queue.pop(0)
        return self.queue.pop(0)
